dpdkstat plugin rework

krzysztofx opened the pull request:
This patch contains rework of existing dpdkstat plugin. Changes are focused on
extracting common dpdk related features into utils file that could be reused by
other DPDK plugins. Also some related structural changes are introduced.

List of changes:
- introduction of new utils_dpdk.c that handles child process spawning,
  and terminating, messaging between parent and child, DPDK configuration
  and initialization and DPDK commands execution
- DPDK EAL specific options put into separate subsection of
  dpdkstat configuration
- documentation update

Those changes are preparatory steps for upstreaming dpdkevent plugin
that is going to collect DPDK ports and logical cores status info.

Signed-off-by: Krzysztof Matczak <krzysztofx.matczak@intel.com>

rubenk commented on docs/BUILD.dpdkstat.md:
```diff
@@ -201,15 +201,19 @@ See also: http://dpdk.org/doc/guides/prog_guide/multi_proc_support.html
  *  Configure collectd with the DPDK shared library:
 
-        ./configure --with-libdpdk=/usr
+        ./configure LIBDPDK_CPPFLAGS="-I/usr/include/dpdk" LIBDPDK_LDFLAGS="-L/usr/lib"
```
Setting these flags isn't needed anymore since 8259168611932117193bcf891b1c063a619d32e3

krzysztofx replied:
Correct me if I'm wrong : you mean that setting these flags is not needed if we set it as above since those are default values, right ?
If so, then what about clarifying that if user needs to set custom header/lib dpdk locations, then may set it via these variables,  exactly like in [8259168](https://github.com/collectd/collectd/commit/8259168611932117193bcf891b1c063a619d32e3) ./configure LIBDPDK_CPPFLAGS="-I/path/to/your/libdpdk_headers" LIBDPDK_LDFLAGS="-L/path/to/your/libdpdk.so -Wl,--no-as-needed"

>

rubenk commented on docs/BUILD.dpdkstat.md:
```diff
@@ -201,15 +201,19 @@ See also: http://dpdk.org/doc/guides/prog_guide/multi_proc_support.html
  *  Run *configure* with the following CFLAGS:
 
-        ./configure --with-libdpdk=/usr CFLAGS=" -lpthread -Wl,--whole-archive -Wl,-ldpdk -Wl,-lm -Wl,-lrt -Wl,-lpcap -Wl,-ldl -Wl,--no-whole-archive"
+        ./configure LIBDPDK_CPPFLAGS="-I/usr/include/dpdk" LIBDPDK_LDFLAGS="-L/usr/lib" CFLAGS=" -lpthread -Wl,--whole-archive -Wl,-ldpdk -Wl,-lm -Wl,-lrt -Wl,-lpcap -Wl,-ldl -Wl,--no-whole-archive"
```
Same here. Unrelated to this PR, but the other libraries are probably not needed too.

krzysztofx replied:
let me verify that for static linking

rubenk replied:
I'm not entirely sure what you mean. We don't support static linking of external libraries into plugins.

krzysztofx replied:
@rubenk 
That chapter is  dedicated to  static linking of DPDK libraries into the plugin. I didn't know that it's not supported in collectd.  What does it mean exactly - you just  recommend  not doing that , or there's some technical issue that will prevent user from running collectd with statically linked plugin  ? 

If it's static linking is not supported, then maybe that chapter should be removed at all, what do you think ? 

Regards 
Krzysztof

rubenk replied:
Collectd explicitly disables static linking for libtool (https://github.com/collectd/collectd/blob/master/configure.ac#L14)

I never tried this ./configure line, if it works for you it's fine with me.

krzysztofx replied:
It turned out that there's no clean and universal solution for static linking of DPDK. It changes from one DPDK version to another. We decided to remove all  related content from documentation.

rubenk commented on docs/BUILD.dpdkstat.md:
```diff
@@ -201,15 +201,19 @@ See also: http://dpdk.org/doc/guides/prog_guide/multi_proc_support.html
-        ./configure --with-libdpdk=/usr CFLAGS=" -lpthread -Wl,--whole-archive -Wl,-ldpdk -Wl,-lm -Wl,-lrt -Wl,-lpcap -Wl,-ldl -Wl,--no-whole-archive"
+        ./configure LIBDPDK_CPPFLAGS="-I/usr/include/dpdk" LIBDPDK_LDFLAGS="-L/usr/lib" CFLAGS=" -lpthread -Wl,--whole-archive -Wl,-ldpdk -Wl,-lm -Wl,-lrt -Wl,-lpcap -Wl,-ldl -Wl,--no-whole-archive"
 
  *  Make sure that dpdk and dpdkstat are enabled in the *configure* output.
```
libdpdk, might be nice to fix this too.

krzysztofx replied:
ok

rubenk commented on src/dpdkstat.c:
```diff
@@ -18,55 +18,37 @@
- * SOFTWARE.
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
```
What caused this rewrapping of the lines? They were already max 80 chars and this seems like unnecessary churn.

krzysztofx replied:
This  rework was branched out long time ago and many things have been changed including license note  reformatting and also list of authors.
I can align license note  in all  files in this PR to base version  if you prefer that.

rubenk replied:
That would be nice, thank you.

rubenk commented on src/dpdkstat.c:
```diff
@@ -18,55 +18,37 @@
  *   Maryam Tahhan <maryam.tahhan@intel.com>
  *   Harry van Haaren <harry.van.haaren@intel.com>
- *   Taras Chornyi <tarasx.chornyi@intel.com>
- */
```
Please don't remove someone elses authorship.

krzysztofx replied:
Ok, that wasn't intentional. see my previous comment. I'll consult that with Taras

rubenk commented on src/dpdkstat.c:
```diff
@@ -88,742 +70,462 @@
-    return err;
+static int dpdk_stats_config(oconfig_item_t *ci) {
+  DPDK_STATS_TRACE();
+  int ret = 0;
```
dead assignment.

krzysztofx commented:
@rubenk 
Thanks for your review. I'll addres your comments ASAP.

regards
Krzysztof

krzysztofx commented on src/dpdkstat.c:
```diff
@@ -88,742 +70,462 @@
+  DPDK_STATS_TRACE();
 
-  char errbuf[ERR_BUF_SIZE];
+  int ret = 0;
```
ok

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  cdtime_t cmd_wait_time;
+
+  pid_t pid;
+#ifdef DPDK_HELPER_USE_PIPES
```
This is always true, so why the #ifdef?

krzysztofx replied:
All logic related with redirecting stdout from child to parent process was wrapped with  DPDK_HELPER_USE_PIPES ifdef just to quickly disable it in future if it's no longer needed. That redirecting was implemented because child process gets important info from  DPDK EAL into it's stdout.  If in some future DPDK version all EAL initialization info will be logged via system logger then we may disable those pipes quickly.
Do you prefer not to contaminate source code with these ifdefs?

rubenk replied:
Yes, I'd prefer if you removed those ifdefs. If they are needed in the future, let's add them then.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+};
+
+#define DPDK_HELPER_TRACE(_name)                                               \
+  DEBUG("%s:%s:%d pid=%lu", _name, __FUNCTION__, __LINE__, (long)getpid())
```
%ld for the pid please.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+    return -EINVAL;
+  }
+
+  memcpy(&phc->eal_config, ec, sizeof(dpdk_eal_config_t));
```
I'd prefer that you call sizeof on the variable, not the type.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+    return -EINVAL;
+  }
+
+  memcpy(ec, &phc->eal_config, sizeof(dpdk_eal_config_t));
```
Same here.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+                "/var/run/.%s_config", child->values[0].value.string);
+      DEBUG("dpdk_common: EAL:File prefix %s", phc->eal_config.file_prefix);
+    }
+  }
```
Please handle errors and unknown config keys in this loop.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  if (ret != 0) {
+    WARNING("dpdk_shm_init: Failed to resize SHM:%s\n",
+            sstrerror(errno, errbuf, sizeof(errbuf)));
+    goto fail_close;
```
I'm not too familiar with SHM, but don't you need to destroy / delete / close it on failure?

krzysztofx replied:
Yes, to be fixed. Additionally  shm_unlink should be called if  shm_open succeeded.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  if (fd < 0) {
+    WARNING("dpdk_shm_init: Failed to open %s as SHM:%s\n", name,
+            sstrerror(errno, errbuf, sizeof(errbuf)));
+    goto fail;
```
I really dislike goto's, especially 2 labels in the same function. Case in point, Coverity found 2 issues in this function with the error handling. I'd prefer that you do the cleanup immediately and return.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  }
+  /*
+   * Close the file descriptor, the shared memory object still exists
+   * and can only be removed by calling shm_unlink().
```
I don't understand this comment. Any reason you don't call shm_unlink here?

krzysztofx replied:
No, I can't call shm_unlink since this is positive scenario, I mean shared memory was successfully initialized. File descriptor is no longer needed for shm operations and can be closed. I guess  that was what comment author wanted to say.
I'll  modify this comment.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  return 0;
+}
+
+inline void *dpdk_helper_priv_get(dpdk_helper_ctx_t *phc) {
```
No explicit inlines please, let the compiler figure it out.

krzysztofx replied:
ok

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+
+inline void *dpdk_helper_priv_get(dpdk_helper_ctx_t *phc) {
+  if (phc)
+    return (void *)phc->priv_data;
```
unnecessary cast.

krzysztofx replied:
ok

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+
+int dpdk_helper_init(const char *name, size_t data_size,
+                     dpdk_helper_ctx_t **pphc) {
+  int err = 0;
```
Initialisation has no effect, please declare err when you need it.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+
+  int ret = 0;
+
+  ret = dpdk_helper_status_check(phc);
```
Please don't write code like this.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+          sstrerror(errno, errbuf, sizeof(errbuf)));
+    sem_destroy(&phc->sema_cmd_start);
+    dpdk_shm_cleanup(name, shm_size, (void *)phc);
+    return -errno;
```
same here.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+    return -1;
+  }
+
+  int pipe0_flags = fcntl(phc->pipes[0], F_GETFL, 0);
```
please check for error.

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+  uint128_t lcore_mask;
+  int err = 0;
+
+  memset(&lcore_mask, 0, sizeof(uint128_t));
```
sizeof(lcore_mask)

rubenk commented on src/utils_dpdk.c:
```diff
@@ -0,0 +1,871 @@
+
+    lcore_mask.low = strtoull_safe(low_str, &err);
+    if (err)
+      goto parse_out;
```
please just get rid of the goto and return immediately.

krzysztofx commented on src/utils_dpdk.h:
```diff
@@ -0,0 +1,97 @@
+#if RTE_VER_MONTH == 7
+#define DPDK_VER_16_07 RTE_VERSION_NUM(16, 7, 0, 16)
+#endif
+#endif
```
Yes. I will remove it

krzysztofx commented:
@rubenk 
I have addressed your comments in the last patch. There are some merge conflicts, so should I rebase to master branch ?

Thanks
Krzysztof

rubenk commented:
Thank you @krzysztofx. Yes, please rebase.

krzysztofx commented:
@rubenk 
Rebase done.  As expected, epel5 builds are  failing due to unsupported socket flags. All other builds passed. 

Thanks
Krzysztof
